include/SampleThumbnail.h

```diff
@@ -0,0 +1,155 @@
+
+#include <QPainter>
+#include <QRect>
+#include <memory>
+#include <QDateTime>
+
+#include "Sample.h"
+#include "lmms_export.h"
+
+namespace lmms {
+
+/**
+   Allows for visualizing sample data.
+
+   On construction, thumbnails will be generated
+   at logarathmic intervals of downsampling. Those cached thumbnails will then be further downsampled on the fly and
+   transformed in various ways to create the desired waveform.
+
+   Given that we are dealing with far less data to generate
+   the visualization however (i.e., we are not reading from original sample data when drawing), this provides a
+   significant performance boost that wouldn't be possible otherwise.
+ */
+class LMMS_EXPORT SampleThumbnail
+{
+public:
+	//! Maximum number of cached sample thumbnails.
+	//! This cache only allows for caching sample thumbnails that have an associated file path.
+	static constexpr auto MaxSampleThumbnailCacheSize = 32;
+
+	struct VisualizeParameters
+	{
+		QRect sampleRect; //!< A rectangle that covers the entire range of samples.
+
+		QRect drawRect; //!< Specifies the location in `sampleRect` where the waveform will be drawn. Equals
+						//!< `sampleRect` when null.
+
+		QRect viewportRect; //!< Clips `drawRect`. Equals `drawRect` when null.
+
+		float amplification = 1.0f; //!< The amount of amplification to apply to the waveform.
+
+		float sampleStart = 0.0f; //!< Where the sample begins for drawing.
+
+		float sampleEnd = 1.0f; //!< Where the sample ends for drawing.
+
+		bool reversed = false; //!< Determines if the waveform is drawn in reverse or not.
+	};
+
+	SampleThumbnail() = default;
+	SampleThumbnail(const Sample& sample);
+	void visualize(const VisualizeParameters& parameters, QPainter& painter) const;
+
+private:
+	class Thumbnail
+	{
+	public:
+		static constexpr auto AggregationPerZoomStep = 10;
+
+		struct Peak
+		{
+			Peak() = default;
+
+			Peak(float min, float max)
+				: min(min)
+				, max(max)
+			{
+			}
+
+			Peak(const SampleFrame& frame)
+				: min(std::min(frame.left(), frame.right()))
+				, max(std::max(frame.left(), frame.right()))
+			{
+			}
+
+			Peak operator+(const Peak& other) const { return Peak(std::min(min, other.min), std::max(max, other.max)); }
+			Peak operator+(const SampleFrame& frame) const { return *this + Peak{frame}; }
+
+			float min = std::numeric_limits<float>::max();
+			float max = -std::numeric_limits<float>::max();
+		};
+
+		Thumbnail() = default;
+		Thumbnail(std::vector<Peak> peaks, double samplesPerPeak);
+		Thumbnail(const SampleFrame* buffer, size_t size, size_t width);
+
+		Thumbnail zoomOut(float factor) const;
+		Thumbnail zoomOut(float factor, size_t from, size_t to) const;
+
+		Peak& operator[](size_t index) { return m_peaks[index]; }
+		const Peak& operator[](size_t index) const { return m_peaks[index]; }
+
+		Peak* peaks() { return m_peaks.data(); }
+		const Peak* peaks() const { return m_peaks.data(); }
+
+		int width() const { return m_peaks.size(); }
```

Comment:
Should this return `std::size_t`?

Reply:
Since the math involved uses `int` quite often (stuff like `std::max(rect->x(), thumbnail->width()`), making `width` a `size_t` is more of an inconvenience than anything else, though one could argue that returning `size_t` is more correct here. Using `std::max<int>` is an option, but I don't like the possible implicit conversion (which had caused bugs in the past relating to sample waveform visualization)